### PR TITLE
Simulator in a web worker

### PR DESCRIPTION
**Pull request by yann300**
This PR:

 - use ethereumjs/vm 6.0.0 and its dependencies.
 - put remix-simulator  in a WebWorker.

### CODE REVIEW

File: apps/remix-ide/src/blockchain/providers/worker-vm.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { Provider } from '@remix-project/remix-simulator'
+import { bigIntToHex } from '@ethereumjs/util'
```
Comment (bunsenstraat) [inline apps/remix-ide/src/blockchain/providers/worker-vm.ts:2]: unused import

File: libs/remix-lib/src/execution/txRunnerWeb3.ts [hunk 2]
```diff
@@ -47,6 +47,9 @@ export class TxRunnerWeb3 {
   }
 
   _sendTransaction (sendTx, tx, pass, callback) {
+    let currentDateTime = new Date();
+    const start = currentDateTime.getTime() / 1000
+    console.log('start', start)
```
Comment (Aniket-Engg) [inline libs/remix-lib/src/execution/txRunnerWeb3.ts:52]: remove log

File: libs/remix-lib/src/execution/txRunnerVM.ts [hunk 3]
```diff
@@ -1,10 +1,21 @@
 'use strict'
+import { RunBlockResult, RunTxResult } from '@ethereumjs/vm'
 import { Transaction, FeeMarketEIP1559Transaction } from '@ethereumjs/tx'
 import { Block } from '@ethereumjs/block'
 import { BN, bufferToHex, Address } from 'ethereumjs-util'
```
Comment (Aniket-Engg) [inline libs/remix-lib/src/execution/txRunnerVM.ts:5]: I think `@ethereumjs/util` and `'ethereumjs-util'` are same

File: libs/remix-lib/src/execution/txRunnerWeb3.ts (hunk 2 above)
Comment (yann300) [reply libs/remix-lib/src/execution/txRunnerWeb3.ts:52]: i am using this log to track performances, i'd rather keep it for the moment

**Comment by yann300** [comment]
I will take care of the following in a next PR:
 - moving to ethereumjs/vm 6.3.0
 - fixing the ethereumjs-util vs @ethereumjs/util dep

File: apps/remix-ide-e2e/src/commands/checkVariableDebug.ts [hunk 4]
```diff
@@ -29,6 +29,7 @@ function checkDebug(browser: NightwatchBrowser, id: string, debugValue: Nightwat
     }, [id], (result) => {
       if (result.value) {
         try {
+          console.log('checkDebug', result.value)
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/commands/checkVariableDebug.ts:32]: can this be removed?

File: apps/remix-ide-e2e/src/tests/remixd.test.ts [hunk 5]
```diff
@@ -311,15 +311,24 @@ function testImportFromRemixd(browser: NightwatchBrowser, callback: VoidFunction
...
+async function spawnRemixd(path: string): Promise<ChildProcess> {
   const remixd = spawn('yarn run remixd', [`-s ${path}`], { cwd: process.cwd(), shell: true, detached: true })
-  remixd.stdout.on('data', function (data) {
-    console.log('stdout: ' + data.toString())
-  })
-  remixd.stderr.on('err', function (data) {
-    console.log('err: ' + data.toString())
+  return new Promise((resolve, reject) => {
+    remixd.stdout.on('data', function (data) {
+      console.log('stdout: ' + data.toString())
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/remixd.test.ts:318]: same here

File: apps/remix-ide-e2e/src/tests/remixd.test.ts [hunk 6]
```diff
@@ -311,15 +311,24 @@ function testImportFromRemixd(browser: NightwatchBrowser, callback: VoidFunction
...
+      if(
+        data.toString().includes('is listening') 
+        || data.toString().includes('There is already a client running')
+        ) {
+        
+        resolve(remixd)
+      }
+    })
+    remixd.stderr.on('err', function (data) {
+      console.log('err: ' + data.toString())
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/remixd.test.ts:328]: this log

File: apps/remix-ide/src/blockchain/providers/worker-vm.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+              cmd: 'sendAsyncResult',
+              error,
+              result,
+              stamp: data.stamp
+            })
+        })
+      } else {
+        self.postMessage({
+          cmd: 'sendAsyncResult',
+          error: 'Provider not instanciated',
```
Comment (Aniket-Engg) [inline apps/remix-ide/src/blockchain/providers/worker-vm.ts:27]: typo in `instanciated `

File: libs/remix-lib/src/execution/txRunnerWeb3.ts [hunk 8]
```diff
@@ -57,6 +60,10 @@ export class TxRunnerWeb3 {
         return new Promise(async (resolve, reject) => {
           const receipt = await tryTillReceiptAvailable(resp, this.getWeb3())
           tx = await tryTillTxAvailable(resp, this.getWeb3())
+          currentDateTime = new Date();
+          const end = currentDateTime.getTime() / 1000
+          console.log('end', end)
+          console.log('stopwatch', end - start)
```
Comment (Aniket-Engg) [inline libs/remix-lib/src/execution/txRunnerWeb3.ts:66]: would you like to keep these two logs too?

File: libs/remix-simulator/src/VmProxy.ts [hunk 9]
```diff
@@ -133,23 +135,21 @@ export class VmProxy {
...
     if (data.to) {
-      try {
-        const storage = await this.vm.stateManager.dumpStorage(data.to)
-        this.storageCache[this.processingHash][tx['to']] = storage
-        this.lastProcessedStorageTxHash[tx['to']] = this.processingHash
-      } catch (e) {
-        console.log(e)
-      }
+      ((processingHash, processingAccount, processingAddress, self) => {
+        self.stateCopy.dumpStorage(processingAccount).then((storage) => {
```
Comment (Aniket-Engg) [inline libs/remix-simulator/src/VmProxy.ts:139]: can we use `await` here? method is already `async`

File: libs/remix-simulator/src/VmProxy.ts [hunk 10]
```diff
@@ -133,23 +135,21 @@ export class VmProxy {
...
     this.storageCache['after_' + this.processingHash] = {}
     if (data.to) {
-      try {
-        const storage = await this.vm.stateManager.dumpStorage(data.to)
-        this.storageCache[this.processingHash][tx['to']] = storage
-        this.lastProcessedStorageTxHash[tx['to']] = this.processingHash
-      } catch (e) {
-        console.log(e)
-      }
+      ((processingHash, processingAccount, processingAddress, self) => {
```
Comment (Aniket-Engg) [inline libs/remix-simulator/src/VmProxy.ts:138]: any specific reason to use anonymous method?

File: libs/remix-simulator/src/VmProxy.ts [hunk 11]
```diff
@@ -191,95 +191,100 @@ export class VmProxy {
...
-      const fnselector = parseInt(fnselectorStrInHex)
-      const fnArgs = ConsoleLogs[fnselector]
-      const iface = new ethers.utils.Interface([`function log${fnArgs} view`])
-      const functionDesc = iface.getFunction(`log${fnArgs}`)
-      const sigHash = iface.getSighash(`log${fnArgs}`)
-      if (fnArgs.includes('uint') && sigHash !== fnselectorStrInHex) {
-        payload = payload.replace(fnselectorStr, sigHash)
-      } else {
-        payload = '0x' + payload
+      if (this.previousDepth > depth && previousopcode) {
```
Comment (Aniket-Engg) [inline libs/remix-simulator/src/VmProxy.ts:215]: `previousopcode ` should be in camelcase

File: libs/remix-simulator/src/VmProxy.ts [hunk 12]
```diff
@@ -191,95 +191,100 @@ export class VmProxy {
...
+        if (step.op === 'CREATE') {
+          this.processingAddress = '(Contract Creation - Step ' + this.processingIndex + ')'
+          this.storageCache[this.processingHash][this.processingAddress] = {}
+        } else {
+          this.processingAddress = normalizeHexAddress(step.stack[step.stack.length - 2])
+          this.processingAddress = toChecksumAddress(this.processingAddress)
+          if (!this.storageCache[this.processingHash][this.processingAddress]) {
+            ((processingHash, processingAddress, self) => {
+                const account = Address.fromString(processingAddress)
+                self.stateCopy.dumpStorage(account).then((storage) => {
```
Comment (Aniket-Engg) [inline libs/remix-simulator/src/VmProxy.ts:269]: same here

File: libs/remix-simulator/src/methods/accounts.ts [hunk 13]
```diff
@@ -1,18 +1,18 @@
 import { BN, privateToAddress, toChecksumAddress, isValidPrivate, Address } from 'ethereumjs-util'
-import Web3 from 'web3'
+const Web3EthAccounts = require('web3-eth-accounts');
```
Comment (Aniket-Engg) [inline libs/remix-simulator/src/methods/accounts.ts:2]: semicolon can be removed

File: libs/remix-simulator/src/vm-context.ts [hunk 14]
```diff
@@ -115,14 +129,18 @@ export class VMContext {
...
     const stateManager = new StateManagerCommonStorageDump()
     const common = new Common({ chain: 'mainnet', hardfork })
-    const vm = new VM({
+    const vm = await VM.create({
       common,
       activatePrecompiles: true,
-      stateManager,
-      allowUnlimitedContractSize: true
+      stateManager
+      // allowUnlimitedContractSize: true
```
Comment (Aniket-Engg) [inline libs/remix-simulator/src/vm-context.ts:143]: can this param be removed?

File: libs/remix-simulator/src/VmProxy.ts (hunk 10 above)
Comment (yann300) [reply libs/remix-simulator/src/VmProxy.ts:138]: yes, the root function is a callback, called multiple times. `self.stateCopy.dumpStorage(account)` take some time to resolve, I want to make sure that there's no interference between multiple call to `dumpStorage`, using enclosure allows a clear separation.

File: libs/remix-lib/src/execution/txRunnerWeb3.ts (hunk 8 above)
Comment (Aniket-Engg) [reply libs/remix-lib/src/execution/txRunnerWeb3.ts:64]: ??

Comment (yann300) [reply libs/remix-lib/src/execution/txRunnerWeb3.ts:64]: I removed the other logs and kept this one.

**Review by Aniket-Engg — APPROVED** [review]

**Pull request merged**